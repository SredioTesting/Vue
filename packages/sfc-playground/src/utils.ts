export function debounce(fn: Function, n = 100) {
  let handle: number | null = null
  return (...args: any[]) => {
    if (handle) window.clearTimeout(handle)
    handle = window.setTimeout(() => {
      fn(...args)
      handle = null
    }, n)
  }
}

// prefer old unicode hacks for backward compatibility
// https://base64.guru/developers/javascript/examples/unicode-strings
export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}

export function atou(base64: string): string {
  return decodeURIComponent(escape(atob(base64)))
}
