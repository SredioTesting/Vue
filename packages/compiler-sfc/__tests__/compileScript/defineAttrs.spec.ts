import { compileSFCScript as compile, assertCode } from '../utils'

describe('defineAttrs()', () => {
  test('basic usage', () => {
    const { content } = compile(`
      <script setup lang="ts">
      const attrs = defineAttrs<{
        bar?: number
      }>()
      </script>
    `)
    assertCode(content)
    expect(content).toMatch(`const attrs = _useAttrs()`)
    expect(content).not.toMatch('defineAttrs')
  })

  test('w/o return value', () => {
    const { content } = compile(`
      <script setup lang="ts">
      defineAttrs<{
        bar?: number
      }>()
      </script>
    `)
    assertCode(content)
    expect(content).not.toMatch('defineAttrs')
    expect(content).not.toMatch(`_useAttrs`)
  })

  test('w/o generic params', () => {
    const { content } = compile(`
      <script setup>
      const attrs = defineAttrs()
      </script>
    `)
    assertCode(content)
    expect(content).toMatch(`const attrs = _useAttrs()`)
    expect(content).not.toMatch('defineAttrs')
  })
})
