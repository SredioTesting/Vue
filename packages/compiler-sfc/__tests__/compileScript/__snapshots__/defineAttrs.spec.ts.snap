// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineAttrs() > basic usage 1`] = `
"import { useAttrs as _useAttrs, defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose: __expose }) {
  __expose();

      const attrs = _useAttrs()
      
return { attrs }
}

})"
`;

exports[`defineAttrs() > w/o generic params 1`] = `
"import { useAttrs as _useAttrs } from 'vue'

export default {
  setup(__props, { expose: __expose }) {
  __expose();

      const attrs = _useAttrs()
      
return { attrs }
}

}"
`;

exports[`defineAttrs() > w/o return value 1`] = `
"import { defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose: __expose }) {
  __expose();

      
      
return {  }
}

})"
`;
